Add functionalization pass in TorchDynamo

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99461

Fixes: https://github.com/pytorch/pytorch/issues/99000

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

tugsbayasgalan committed:
Add functionalization pass in TorchDynamo

[ghstack-poisoned]

bdhirsh commented on torch/_dynamo/config.py:
```diff
@@ -73,6 +73,9 @@
 guard_nn_modules = False
 
+# Optionally functionalize the produced export graph module
+functionalize = False
```
side note - @anijain2305 what's our stance towards BC on torch._dynamo.export? Are we ok with removing kwargs later?

anijain2305 replied:
All the options seem equally bad to me. Imo, export is going through a heavy refactoring, and I would wait for aot_export etc to be finalized before we set the API in stone.

bdhirsh commented on torch/_dynamo/eval_frame.py:
```diff
@@ -166,6 +166,31 @@ def enable_dynamic(enable: bool = True, export: bool = False):
+        torch._enable_functionalization(reapply_views=True)
+        yield pytree.tree_map(to_fun, args)
+    finally:
+        torch._disable_functionalization()
```
I think this will still cause problems, if your `Interpreter(graph).run()` call lower down raises an exception. You need that to be wrapped by the try/finally to make sure that functionalization is always properly turned off afterwards.

One way would be to change this context manager into a decorator that wraps your interpreter call. Or you can just not bother with the decorator here, and manually try/finally around the interpreter call.

tugsbayasgalan replied:
Good catch! I will just do plain try/finally.

bdhirsh commented:
Otherwise look ok to me!

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh reviewed (approved):
lgtm!

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45106409).

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45106409).

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45106409).

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45106409).

zou3519 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -670,6 +671,9 @@ def export(
         If decomposition_table is specified, the options are "symbolic" or "fake". Default is "real".
 
+        functionalize (bool): If True, the resulting aten graph module will be functional. You will need to
+        set aten_graph=True to see the effect.
```
nit: specify default

zou3519 commented on test/dynamo/test_export.py:
```diff
@@ -2422,6 +2422,115 @@ def f_correct(x):
+    # it specializes predicates now. we need to fix it.
+    @unittest.expectedFailure
+    @config.patch(dynamic_shapes=True)
+    def test_functionalize_cond(self):
+        def foo(x):
+            def true_true_fn(x):
+                return x.sum() + 6
+
+            def true_false_fn(x):
+                return x.sum() + 9
+
+            def true_fn(x):
+                return cond(x.shape[0] > 6, true_true_fn, true_false_fn, [x])
+
+            def false_fn(x):
+                return x.sum() - 1
+
+            return cond(x.shape[0] > 5, true_fn, false_fn, [x])
+
+        example_inputs = (torch.ones(5, 2, 3),)
+        gm, _ = torch._dynamo.export(
+            foo,
+            *example_inputs,
+            aten_graph=True,
+            tracing_mode="symbolic",
+            functionalize=True,
+        )
+        self.assertEqual(gm(torch.ones(7, 2, 3)), foo(torch.ones(7, 2, 3)))
```
All of these operations are already functional, we should test some functions that have mutation in them.

tugsbayasgalan replied:
Added some tests in test/functorch/test_control_flow.py

zou3519 commented on functorch/experimental/_cond.py:
```diff
@@ -193,18 +200,41 @@ def _has_potential_branch_input_alias(branch, inputs):
+def cond_func(pred, true_fn, false_fn, inputs):
+    unwrapped_inputs = _unwrap_all_tensors_from_functional(inputs, reapply_views=torch._C._functionalization_reapply_views_tls())
+    unwrapped_pred = _unwrap_all_tensors_from_functional(pred, reapply_views=torch._C._functionalization_reapply_views_tls())
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Functionalize))
```
These type of guards need to be wrapped in try-finally, otherwise if an exception happens they won't be deleted (the exception traceback may hold onto the object).

i.e.
```py
guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Functionalize))
try:
    ...
finally:
    del guard
```

tugsbayasgalan replied:
Yep did.

zou3519 commented on functorch/experimental/_cond.py:
```diff
@@ -193,18 +200,41 @@ def _has_potential_branch_input_alias(branch, inputs):
+            raise UnsupportedAliasMutationException("One of torch.cond branch "
+                                                    "might be aliasing the input!")
 
+    return _wrap_all_tensors_to_functional(cond(unwrapped_pred, true_fn, false_fn, unwrapped_inputs), level=0)
```
I don't think this actually works -- we disabled the functionalization key, so the branches will not get run through with functional tensors. I suggested some test cases below, it doesn't look like we are actually testing any conds with in-place ops in the branch functions.

My take is we still need to apply functionalize(true_fn) and functionalize(false_fn) here

zou3519 reviewed (commented):
Need to run for the day, but I'm not sure the cond functionalize rule actually works

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45106409).

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45106409).

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45106409).

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45106409).

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -750,6 +754,12 @@ def export(
+    if functionalize and not aten_graph:
+        raise UserError(
+            UserErrorType.ANTI_PATTERN,
+            "TorchDynamo won't functionalize non-aten graphs. Please set `functionalize` to true",
```
```suggestion
            "TorchDynamo won't functionalize non-aten graphs. Please set `aten_graph` to true",
```

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -904,10 +914,40 @@ def run_node(self, n):
             with torch.fx.traceback.preserve_node_meta():
-                return torch.fx.Interpreter(graph).run(*args)
+                if functionalize:
+                    torch._enable_functionalization(reapply_views=False)
```
I think we concluded to allow views in the graph, should we have reapply_views=True?

tugsbayasgalan replied:
Oh lol you are right. We will replace them with view_copy later as pass

gmagogsfm commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1033,10 +1073,6 @@ def argument_names(f: Callable[..., Any], *args, **kwargs) -> List[str]:
 
     new_graph.recompile()
 
-    # TODO remove this once Executorch uses proper functionalization
```
Love this deletion!

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add functionalization pass in TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/99000

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D45106409](https://our.internmc.facebook.com/intern/diff/D45106409)

[ghstack-poisoned]
